File: shared/web-sdk/src/connectors/sessions-connector.ts
import type { SessionRow } from "@sentient/protocol";
import type { Connector, SentientSDKInternal } from "../connector-types.ts";
import { createLogger } from "../logger.ts";
import type { SessionsListResult, SessionsRest } from "../sessions-rest.ts";

const log = createLogger(["sentient", "sdk", "connectors", "sessions"]);

const DEFAULT_TIMEOUT_MS = 5000;

export type SessionsChangeEvent =
  | { kind: "created"; sessionId: string; title?: string; ts: number }
  | { kind: "switched"; sessionId: string; title?: string; ts: number }
  /** No session yet. `draftKey` is the opaque handle the gateway minted for
   *  this draft: the client holds it as its current pointer, re-presents it on
   *  `session.configure`, and the gateway spends it as the mint key when the
   *  first message arrives. A draft is NOT a session — it has no row and must
   *  never appear in the session list. `deletionId` is SDK-local correlation
   *  for a draft handshake caused by an unavailable session. */
  | { kind: "draft"; draftKey: string; ts: number; deletionId?: string }
  | { kind: "deleted"; sessionId: string; deletionId?: string }
  /** A reconnect presented a session the caller's store no longer has. This
   *  is local recovery metadata, not a server delete command. */
  | { kind: "unavailable"; sessionId: string; deletionId: string }
  | { kind: "renamed"; sessionId: string; title: string };

export interface SessionsConnectorConfig {
  readonly rest: SessionsRest;
  readonly timeoutMs?: number;
}

const newId = (): string => globalThis.crypto.randomUUID();

export class SessionsConnector implements Connector {
  readonly capability = "sessions";
  readonly kind = "status" as const;

  private readonly rest: SessionsRest;
  private readonly timeoutMs: number;
  private listeners: Set<(e: SessionsChangeEvent) => void> = new Set();
  private unsubs: (() => void)[] = [];
  private send: (msg: Record<string, unknown>) => void = () => {};
  private pendingSwitch: { reject(error: Error): void } | null = null;
  private readonly deletedSessionIds = new Set<string>();
  /** Current connection's exact old session, paired with its no-request draft handshake. */
  private pendingUnboundSessionId: string | null = null;

  constructor(cfg: SessionsConnectorConfig) {
    this.rest = cfg.rest;
    this.timeoutMs = cfg.timeoutMs ?? DEFAULT_TIMEOUT_MS;
  }

  attach(sdk: SentientSDKInternal): void {
    this.send = (msg) => sdk.send(msg);
    this.pendingUnboundSessionId = null;

    const dispatchLifecycle = (kind: "created" | "switched") => (raw: unknown) => {
      const m = raw as { sessionId: string; title?: string; ts: number };
      this.dispatch(
        m.title === undefined
          ? { kind, sessionId: m.sessionId, ts: m.ts }
          : { kind, sessionId: m.sessionId, title: m.title, ts: m.ts },
      );
    };
    this.unsubs.push(sdk.onMessage("session.created", dispatchLifecycle("created")));
    this.unsubs.push(sdk.onMessage("session.switched", dispatchLifecycle("switched")));
    this.unsubs.push(
      sdk.onMessage("sessions.deleted", (raw: unknown) => {
        const message = raw as { sessionId?: unknown };
        if (typeof message.sessionId !== "string" || message.sessionId.length === 0) return;
        // The server sends this immediately before the draft handshake for a
        // connection that was attached to this exact deleted session. The SDK
        // supplies the paired old id below; normal New Chat has requestId and
        // never receives this correlation.
        this.dispatchDeleted(message.sessionId, message.sessionId);
      }),
    );
    this.unsubs.push(
      sdk.onMessage("session.refused", (raw: unknown) => {
        const message = raw as { sessionId?: unknown };
        if (typeof message.sessionId !== "string" || message.sessionId.length === 0) return;
        this.pendingUnboundSessionId = message.sessionId;
        this.dispatch({ kind: "unavailable", sessionId: message.sessionId, deletionId: message.sessionId });
      }),
    );
    this.unsubs.push(
      sdk.onMessage("session.unbound", (raw: unknown) => {
        const message = raw as { sessionId?: unknown };
        if (typeof message.sessionId !== "string" || message.sessionId.length === 0) return;
        this.pendingUnboundSessionId = message.sessionId;
        // `sessions.deleted` is normally first and already carries this
        // identity. If it was lost, this event still recovers the boundary.
        if (!this.deletedSessionIds.has(message.sessionId)) {
          this.dispatchDeleted(message.sessionId, message.sessionId);
        }
      }),
    );
    this.unsubs.push(
      sdk.onMessage("sessions.error", (raw: unknown) => {
        const message = raw as { requestId?: unknown; code?: unknown };
        const pending = this.pendingSwitch;
        // conversation.activate is the only session command whose errors have
        // no requestId. Request-bound errors belong to session.new or another
        // command and must not reject an unrelated activation.
        if (pending === null || message.requestId !== undefined || typeof message.code !== "string") return;
        pending.reject(
          new Error(
            message.code === "not_found" || message.code === "forbidden"
              ? "conversation unavailable"
              : "could not open conversation",
          ),
        );
      }),
    );
    this.unsubs.push(
      sdk.onMessage("session.draft", (raw: unknown) => {
        const m = raw as { draftKey?: unknown; ts?: unknown; requestId?: unknown };
        if (typeof m.draftKey !== "string" || typeof m.ts !== "number") return;
        const deletionId = typeof m.requestId === "string" ? undefined : this.pendingUnboundSessionId;
        if (deletionId !== undefined) this.pendingUnboundSessionId = null;
        this.dispatch({
          kind: "draft",
          draftKey: m.draftKey,
          ts: m.ts,
          ...(deletionId === null || deletionId === undefined ? {} : { deletionId }),
        });
      }),
    );
    // The gateway's own titling push. Folded into the `renamed` event rather
    // than given its own kind: every consumer's behaviour is identical — put
    // this title on this row — and the provenance guard that decides WHETHER a
    // generated title may land is enforced in the store (compare-and-set on
    // `title_provenance`), never re-litigated on the client.
    this.unsubs.push(
      sdk.onMessage("session.title", (raw: unknown) => {
        const m = raw as { sessionId: string; title: string };
        this.dispatch({ kind: "renamed", sessionId: m.sessionId, title: m.title });
      }),
    );
  }

  detach(): void {
    for (const u of this.unsubs) u();
    this.unsubs = [];
    this.pendingUnboundSessionId = null;
    this.pendingSwitch?.reject(new Error("session connection closed"));
    // NOTE: do NOT clear `this.listeners` — those are user-registered
    // handlers (e.g. the webui's use-sessions hook subscribes once at
    // create), and they must survive WS reconnect. The SDK calls
    // detach()→attach() on every reconnect; clearing listeners here
    // silently drops every subscription so live session events after the
    // first reconnect never reach the UI. Only `unsubs` (internal
    // sdk.onMessage bindings) are reset.
  }

  private dispatch(e: SessionsChangeEvent): void {
    for (const l of this.listeners) l(e);
  }

  private dispatchDeleted(sessionId: string, deletionId?: string): void {
    // DELETE returns before its broadcast echo necessarily arrives. Treat REST
    // and WS paths as one idempotent lifecycle event while retaining the exact
    // session identity needed to pair its draft handshake.
    if (this.deletedSessionIds.has(sessionId)) return;
    this.deletedSessionIds.add(sessionId);
    this.dispatch({ kind: "deleted", sessionId, ...(deletionId ? { deletionId } : {}) });
  }

  // ---------------------------------------------------------------------------
  // REST-backed query methods
  // ---------------------------------------------------------------------------

  list(opts: { limit?: number; offset?: number } = {}): Promise<SessionsListResult> {
    return this.rest.list(opts);
  }

  search(q: string, limit = 20): Promise<SessionRow[]> {
    return this.rest.search(q, limit);
  }

  async delete(sessionId: string): Promise<void> {
    await this.rest.delete(sessionId);
    this.dispatchDeleted(sessionId);
  }

  async rename(sessionId: string, title: string): Promise<void> {
    await this.rest.rename(sessionId, title);
    this.dispatch({ kind: "renamed", sessionId, title });
  }

  // ---------------------------------------------------------------------------
  // WS-backed session lifecycle methods
  // ---------------------------------------------------------------------------

  /**
   * Activate a conversation. Sends `conversation.activate` over WS and
   * resolves when the gateway broadcasts `session.switched`.
   */
  switchTo(sessionId: string): Promise<void> {
    if (this.pendingSwitch !== null) return Promise.reject(new Error("session switch already pending"));
    return new Promise((resolve, reject) => {
      const cleanup = (): void => {
        clearTimeout(timer);
        this.listeners.delete(onSwitched);
        if (this.pendingSwitch === pending) this.pendingSwitch = null;
      };
      const settle = (action: () => void): void => {
        cleanup();
        action();
      };
      const onSwitched = (e: SessionsChangeEvent): void => {
        if (e.kind === "switched" && e.sessionId === sessionId) settle(resolve);
      };
      const pending = { reject: (error: Error) => settle(() => reject(error)) };
      this.pendingSwitch = pending;
      this.listeners.add(onSwitched);
      const timer = setTimeout(() => {
        log.warn("switchTo.timeout", { reason: "timeout waiting for session.switched", sessionId });
        pending.reject(new Error("timeout waiting for session.switched"));
      }, this.timeoutMs);
      this.send({ type: "conversation.activate", sessionId });
    });
  }

  /**
   * The "+" button: abandon the current session and start a draft.
   *
   * Resolves with the DRAFT KEY, not a session id — pressing "+" mints
   * nothing, so ten presses leave the session list unchanged (spec §4.2). The
   * real id arrives later, as a `created` event, when the first message
   * allocates it.
   *
   * `intent: "explicit"` is what separates this from the `session.new` a client
   * fires on launch. Without it the gateway cannot tell "the person asked for a
   * new chat" from "the app started with no route id" and would abandon the
   * bound session on every launch.
   */
  newChat(): Promise<{ draftKey: string }> {
    return new Promise((resolve, reject) => {
      const onAnswer = (e: SessionsChangeEvent): void => {
        if (e.kind !== "draft" || e.deletionId !== undefined) return;
        cleanup();
        resolve({ draftKey: e.draftKey });
      };
      this.listeners.add(onAnswer);
      const timer = setTimeout(() => {
        cleanup();
        log.warn("newChat.timeout", { reason: "timeout waiting for session.draft" });
        reject(new Error("timeout waiting for session.draft"));
      }, this.timeoutMs);
      const cleanup = (): void => {
        clearTimeout(timer);
        this.listeners.delete(onAnswer);
      };
      this.send({ type: "session.new", requestId: newId(), intent: "explicit" });
    });
  }

  onSessionsChanged(fn: (e: SessionsChangeEvent) => void): () => void {
    this.listeners.add(fn);
    return () => {
      this.listeners.delete(fn);
    };
  }
}

File: shared/mobile-sdk/src/commonMain/kotlin/io/sentient/mobilesdk/connectors/SessionsConnector.kt
// ---------------------------------------------------------------------------
// SessionsConnector — session lifecycle surface for the SDK.
//
// Mirrors web-sdk's sessions-connector.ts post-Task-2.5 shape:
//   capability = "sessions"  (status observer + lifecycle frames)
//
// Query/mutation operations (list/search/delete/rename) are now REST
// (SessionsHttpClient) — the WS query RPCs were removed in protocol Task 2.1.
// Only session-lifecycle WS frames remain here.
//
// LIFECYCLE frame patterns:
//   switchTo / sendSwitch — send conversation.activate; resolve on session.switched
//   newChat / sendNew    — send session.new; resolve on session.created broadcast
//
// Broadcasts (session.created, session.switched, sessions.deleted, sessions.renamed)
// still fan out to onSessionsChanged listeners for live UI.
//
// REST ops (list/search/delete/rename) delegate to the injected
// [SessionsHttpClient] and complete inline — no requestId correlation needed.
//
// KMP-specific:
//   - timeout uses suspend + withTimeout + virtual time for tests.
//   - On timeout the suspend fn throws [SessionsTimeoutException].
//
// Threading: the orchestrator routes frames on its dispatcher; the pending maps +
// listeners are owned here. detach() does NOT clear listeners — they survive
// reconnect. reset() fails pending requests on disconnect.
// ---------------------------------------------------------------------------
package io.sentient.mobilesdk.connectors

import io.sentient.mobilesdk.log.createLogger
import io.sentient.mobilesdk.protocol.ClientMessage
import io.sentient.mobilesdk.protocol.ConversationFeedItem
import io.sentient.mobilesdk.protocol.ServerMessage
import io.sentient.mobilesdk.protocol.SessionRow
import io.sentient.mobilesdk.sessions.SessionsHttpClient
import io.sentient.mobilesdk.util.Clock
import kotlinx.coroutines.CompletableDeferred
import kotlinx.coroutines.TimeoutCancellationException
import kotlinx.coroutines.withTimeout

/** Result of [SessionsConnector.list]. Mirrors web-sdk's list result shape. */
data class SessionsListPage(
    val items: List<SessionRow>,
    val total: Int,
    val hasMore: Boolean,
)

/** A session lifecycle change, fanned out to onSessionsChanged listeners. */
sealed class SessionsChangeEvent {
    data class Created(val sessionId: String, val title: String?, val ts: Long) : SessionsChangeEvent()
    data class Switched(val sessionId: String, val title: String?, val ts: Long) : SessionsChangeEvent()

    /** No session yet — [draftKey] is the handle this client holds until its
     *  first message mints one. A draft has no row and must never be listed. */
    data class Draft(val draftKey: String, val ts: Long) : SessionsChangeEvent()
    data class Deleted(val sessionId: String) : SessionsChangeEvent()
    data class Renamed(val sessionId: String, val title: String) : SessionsChangeEvent()
}

/** Thrown when a session lifecycle request times out. */
class SessionsTimeoutException(val frameType: String) :
    Exception("timeout waiting for $frameType")

/** Thrown when the gateway returns a sessions.error for a lifecycle request. */
class SessionsRequestException(val code: String, override val message: String) :
    Exception("$code: $message")

private const val DEFAULT_TIMEOUT_MS = 5_000L
private const val DEFAULT_MINT_DEBOUNCE_MS = 3_000L

class SessionsConnector(
    private val send: (ClientMessage) -> Unit,
    private val newId: () -> String,
    private val clock: Clock,
    private val httpClient: SessionsHttpClient? = null,
    private val timeoutMs: Long = DEFAULT_TIMEOUT_MS,
    private val mintDebounceMs: Long = DEFAULT_MINT_DEBOUNCE_MS,
) : Connector {
    override val capability: String = CAPABILITY

    private val log = createLogger("connector", "sessions")

    // Connection-scoped debounce. An explicit preparation is retained across a
    // transport edge so a pre-READY request can be retried exactly once.
    private var lastMintAtMs: Long? = null
    private var pendingMintIntent: String? = null

    // Broadcast-correlated waiters (switchTo / newChat).
    private val createdWaiters = mutableListOf<CompletableDeferred<String>>()
    private val switchedWaiters = mutableMapOf<String, CompletableDeferred<Unit>>()

    private val listeners = mutableSetOf<(SessionsChangeEvent) -> Unit>()

    override fun handle(msg: ServerMessage) {
        when (msg) {
            is ServerMessage.SessionsError -> onSessionsError(msg)
            is ServerMessage.SessionsDeleted -> dispatch(SessionsChangeEvent.Deleted(msg.sessionId))
            is ServerMessage.SessionsRenamed -> dispatch(SessionsChangeEvent.Renamed(msg.sessionId, msg.title))
            is ServerMessage.SessionCreated -> onCreated(msg)
            is ServerMessage.SessionSwitched -> onSwitched(msg)
            is ServerMessage.SessionDraft -> onDraft(msg)
            // The gateway's own titling push. Fanned out as Renamed because
            // every consumer does the same thing with it — put this title on
            // this row — and the guard that decides whether a generated title
            // may land at all is enforced in the gateway's store, not here.
            is ServerMessage.SessionTitle -> dispatch(SessionsChangeEvent.Renamed(msg.sessionId, msg.title))
            else -> Unit // not owned by this connector
        }
    }

    private fun onSessionsError(msg: ServerMessage.SessionsError) {
        // Only the switched-waiter path can receive a sessions.error now; reject it.
        // requestId is nullable — conversation.activate errors have none; in that case
        // no waiter exists (fire-and-forget) so the remove is a safe no-op.
        val waiter = switchedWaiters.remove(msg.requestId)
        if (waiter != null) {
            log.warn("switch.error", mapOf("requestId" to msg.requestId, "code" to msg.code))
            waiter.completeExceptionally(SessionsRequestException(msg.code, msg.message))
        }
    }

    private fun onCreated(msg: ServerMessage.SessionCreated) {
        lastMintAtMs = null
        pendingMintIntent = null
        val waiters = createdWaiters.toList()
        createdWaiters.clear()
        for (w in waiters) w.complete(msg.sessionId)
        dispatch(SessionsChangeEvent.Created(msg.sessionId, msg.title, msg.ts))
    }

    /**
     * A draft answers a pending mint exactly as a created session does: the
     * "+" tap is complete, and the key is what the client anchors on so its
     * outbound queue drains. Clearing [lastMintAtMs] here matters — the
     * debounce must not stay armed against a mint that was already answered.
     */
    private fun onDraft(msg: ServerMessage.SessionDraft) {
        lastMintAtMs = null
        pendingMintIntent = null
        val waiters = createdWaiters.toList()
        createdWaiters.clear()
        for (w in waiters) w.complete(msg.draftKey)
        dispatch(SessionsChangeEvent.Draft(msg.draftKey, msg.ts))
    }

    private fun onSwitched(msg: ServerMessage.SessionSwitched) {
        switchedWaiters.remove(msg.sessionId)?.complete(Unit)
        dispatch(SessionsChangeEvent.Switched(msg.sessionId, msg.title, msg.ts))
    }

    private fun dispatch(event: SessionsChangeEvent) {
        for (l in listeners.toList()) l(event)
    }

    // ── REST-backed query ops ──────────────────────────────────────────────────

    /** Page the session list via REST GET /api/v1/sessions. */
    suspend fun list(limit: Int, offset: Int): SessionsListPage {
        log.debug("list", mapOf("limit" to limit, "offset" to offset))
        val items = httpClient?.list(limit, offset) ?: emptyList()
        return SessionsListPage(items = items, total = items.size, hasMore = false)
    }

    /** Search sessions via REST GET /api/v1/sessions/search. */
    suspend fun search(q: String, limit: Int = DEFAULT_SEARCH_LIMIT): List<SessionRow> {
        log.debug("search", mapOf("qLen" to q.length))
        return httpClient?.search(q, limit) ?: emptyList()
    }

    /**
     * Delete a session via REST DELETE /api/v1/sessions/:id.
     * Fans out [SessionsChangeEvent.Deleted] only when the REST call returns true (2xx).
     * On failure, logs a warn and does not fan out — the next list refresh shows reality.
     */
    suspend fun delete(sessionId: String) {
        log.debug("delete", mapOf("sessionId" to sessionId))
        val ok = httpClient?.delete(sessionId) ?: true
        if (ok) {
            dispatch(SessionsChangeEvent.Deleted(sessionId))
        } else {
            log.warn("delete.no-fanout", mapOf("sessionId" to sessionId, "reason" to "REST call failed"))
        }
    }

    /**
     * Rename a session via REST PATCH /api/v1/sessions/:id.
     * Fans out [SessionsChangeEvent.Renamed] only when the REST call returns true (2xx).
     * On failure, logs a warn and does not fan out — the next list refresh shows reality.
     */
    suspend fun rename(sessionId: String, title: String) {
        log.debug("rename", mapOf("sessionId" to sessionId))
        val ok = httpClient?.rename(sessionId, title) ?: true
        if (ok) {
            dispatch(SessionsChangeEvent.Renamed(sessionId, title))
        } else {
            log.warn("rename.no-fanout", mapOf("sessionId" to sessionId, "reason" to "REST call failed"))
        }
    }

    // ── WS lifecycle ops ──────────────────────────────────────────────────────

    /** Send conversation.activate and await the session.switched broadcast. */
    suspend fun switchTo(sessionId: String) {
        val deferred = CompletableDeferred<Unit>()
        switchedWaiters[sessionId] = deferred
        val id = newId()
        log.info("switchTo", mapOf("sessionId" to sessionId, "requestId" to id))
        send(ClientMessage.ConversationActivate(sessionId = sessionId))
        try {
            withTimeout(timeoutMs) { deferred.await() }
        } catch (e: TimeoutCancellationException) {
            switchedWaiters.remove(sessionId)
            log.warn("timeout", mapOf("frame" to "session.switched", "sessionId" to sessionId))
            throw SessionsTimeoutException("session.switched")
        }
    }

    /**
     * Send an EXPLICIT session.new and await the answer — the id on
     * `session.created` when the connection was bound, or the draft key on
     * `session.draft` when the gateway unbound it and there is nothing minted
     * yet.
     */
    suspend fun newChat(): String {
        val deferred = CompletableDeferred<String>()
        createdWaiters.add(deferred)
        val id = newId()
        log.info("newChat", mapOf("requestId" to id))
        send(ClientMessage.SessionNew(requestId = id, intent = INTENT_EXPLICIT))
        return try {
            withTimeout(timeoutMs) { deferred.await() }
        } catch (e: TimeoutCancellationException) {
            createdWaiters.remove(deferred)
            log.warn("timeout", mapOf("frame" to "session.created"))
            throw SessionsTimeoutException("session.created")
        }
    }

    // ── fire-and-forget ops (A2) ──────────────────────────────────────────────

    /** Legacy compatibility operation. Its omitted intent preserves gateway
     * reattachment semantics for non-mobile callers. */
    fun sendNew() = sendNewWithIntent(null)

    /** Prepare an intentionally fresh boundary without blocking the composer. */
    fun startFreshChat() = sendNewWithIntent(INTENT_EXPLICIT)

    private fun sendNewWithIntent(intent: String?) {
        val now = clock.nowMs()
        val inFlight = lastMintAtMs
        if (inFlight != null && now - inFlight < mintDebounceMs) {
            log.info("sendNew.debounced", mapOf("sinceMs" to (now - inFlight)))
            return
        }
        lastMintAtMs = now
        pendingMintIntent = intent
        val id = newId()
        log.info("sendNew", mapOf("requestId" to id, "explicit" to (intent == INTENT_EXPLICIT)))
        send(ClientMessage.SessionNew(requestId = id, intent = intent))
    }

    /**
     * True when a sendNew was attempted but session.created has not yet confirmed it.
     * Used by the SDK to detect a dropped pre-READY mint and re-fire on the READY
     * rising edge (the transport was not open when the original sendNew fired).
     * SDK-internal — not part of the public connector surface.
     */
    internal fun hasPendingMint(): Boolean = lastMintAtMs != null

    /**
     * Clear the mint debounce clock and re-send a session.new. Used exclusively by
     * the SDK on the first-connect READY rising edge when [hasPendingMint] is true:
     * the original sendNew was dropped (null transport), so we retry unconditionally.
     * SDK-internal — not part of the public connector surface.
     */
    internal fun retryPendingMint() {
        val intent = pendingMintIntent
        lastMintAtMs = null
        sendNewWithIntent(intent)
    }

    /** Fire-and-forget switch via conversation.activate. Always sends — no debounce. */
    fun sendSwitch(sessionId: String) {
        log.info("sendSwitch", mapOf("sessionId" to sessionId))
        send(ClientMessage.ConversationActivate(sessionId = sessionId))
    }

    /**
     * Register a sessions-change listener. Returns an unsubscribe fn.
     * Listeners survive detach()/reconnect.
     */
    fun onSessionsChanged(fn: (SessionsChangeEvent) -> Unit): () -> Unit {
        listeners.add(fn)
        return { listeners.remove(fn) }
    }

    /**
     * Fail all in-flight lifecycle requests and clear connection-scoped state.
     * Does NOT clear listeners — they survive WS reconnect.
     * Call on disconnect.
     */
    fun reset() {
        log.info("reset", mapOf("switched" to switchedWaiters.size, "created" to createdWaiters.size))
        // Keep an explicit preparation armed across a transport edge. Implicit
        // compatibility mints remain connection-scoped and are dropped as before.
        if (pendingMintIntent != INTENT_EXPLICIT) {
            lastMintAtMs = null
            pendingMintIntent = null
        }
        val err = SessionsTimeoutException("connector reset")
        for (d in createdWaiters) d.completeExceptionally(err)
        createdWaiters.clear()
        for (d in switchedWaiters.values) d.completeExceptionally(err)
        switchedWaiters.clear()
    }

    companion object {
        const val CAPABILITY: String = "sessions"
        private const val DEFAULT_SEARCH_LIMIT = 20

        /** `session.new.intent` — a person pressing "+", as opposed to the app
         *  launching with no route id (the gateway's default, "implicit"). */
        private const val INTENT_EXPLICIT = "explicit"
    }
}
